torch/_export/graph_module.py

```diff
@@ -224,61 +193,131 @@ def op_to_str(op):
+            else {}
+        )
+
+        # group by input id
+        input_shape_constraints_by_tensor_id = defaultdict(list)
+        for constraint in input_shape_constraints:
+            input_shape_constraints_by_tensor_id[constraint["t_id"]].append(
+                (constraint["dim"], constraint["min"], constraint["max"])
+            )
+
+        input_shape_constraints_by_src_name: Dict[str, List[Tuple[int, ConstraintExpr, ConstraintExpr]]] = {}
+        input_name_to_example_inputs: Dict[str, Any] = {}
+        if example_inputs is not None:
+            input_tracker = 0
+            for node in graph_module.graph.nodes:
+                if node.op == "placeholder":
+                    example_input = example_inputs[input_tracker]
+                    if id(example_input) in input_shape_constraints_by_tensor_id:
+                        input_shape_constraints_by_src_name[node.name] = input_shape_constraints_by_tensor_id[id(example_input)]
+                    input_name_to_example_inputs[node.name] = example_input
+                    input_tracker += 1
+
+        self._input_shape_constraints = input_shape_constraints_by_src_name
+        self._input_name_to_example_inputs = input_name_to_example_inputs
+
+    def __call__(self, *args: Any):
+        return self.forward(*args)
+
+    def forward(self, *args: Any) -> Any:
+        in_spec = self.call_spec.in_spec if self.call_spec is not None else None
+        out_spec = self.call_spec.out_spec if self.call_spec is not None else None
+
+        if in_spec is not None:
+            try:
+                args = fx_pytree.tree_flatten_spec(args, in_spec)  # type: ignore[assignment]
+            except Exception as e:
+                raise error.InternalError("The in_spec is not correctly maintained.") from e
+
+        with torch.fx.traceback.preserve_node_meta(), torch.no_grad():
+            res = torch.fx.Interpreter(self.graph_module).run(*args, enable_io_processing=False)
+
+        if out_spec is not None:
+            try:
+                mutation = self.graph_signature.buffers_to_mutate if self.graph_signature is not None else None
+                num_mutated = len(mutation) if mutation is not None else 0
+                res = pytree.tree_unflatten(res[num_mutated:], out_spec)
+                return res
+            except Exception as e:
+                raise error.InternalError("The out_spec is not correctly maintained.") from e
+        return res
+
+    def __str__(self) -> str:
+        graph_module = self.graph_module.print_readable(print_output=False).replace("\n", "\n    ")
+        string = (
+            "ExportedProgram:\n"
+            f"    {graph_module}\n"
+            f"Graph Signature: {self.graph_signature}\n"
+            f"Symbol to range: {self.symbol_to_range}\n"
+        )
+        return string
+
+    @property
+    def graph(self):
+        return self.graph_module.graph
+
+    def transform(self, *passes: PassType) -> "ExportedProgram":
```

Comment:
Hmm maybe we should create a totally new ExportedProgram object? What do you think?

Reply:
Will copying over the state_dict each time be costly?